### PR TITLE
JS: model property projection calls

### PR DESCRIPTION
**Pull request by esben-semmle**
This PR adds a taint step for property reads done through calls that projects selected properties of an object.

The property projections that have been modeled have slightly different features. So I have adopted @asger-semmle's extensible architecture proposal immediately. We may want to discuss how we generally document the intended usage of the architecture, I have simply added an explanation to the module docstring for now.

This PR only makes use of the object-argument in the models.

For the record, the libraries support various features for the selector-argument:

-   access path strings like `'a.b[42].c'`
-   access path arrays like `['a', 'b', 42, 'c']`
-   search strings like `'a.\*.c'`
-   property predicate functions

We can consider implementing support for these once our general heap treatment is better.

An evaluation with the security suite on big-apps.slugs flags nothing new, and it does not have a significant performance impact ([internal link](https://git.semmle.com/gist/esben/b53bc0d42808e84f8e1c7a5380cddc24)).

### CODE REVIEW

File: javascript/ql/src/semmle/javascript/frameworks/PropertyProjection.qll [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+          name = "pickBy" |
+          callee = DataFlow::moduleMember("ramda", name) and
+          objectIndex = 1 and
+          selectorIndex = 0
+        )
+        or
+        (
+          callee = DataFlow::moduleMember("dotty", "search") and
+          objectIndex = 0 and
+selectorIndex = 1
```
Comment (xiemaisi) [inline javascript/ql/src/semmle/javascript/frameworks/PropertyProjection.qll]: Indentation.

File: javascript/ql/src/semmle/javascript/frameworks/PropertyProjection.qll [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+ * Provides classes for modelling property projection functions.
+ *
+ * Subclass `PropertyProjection` to refine the behavior of the analysis on existing property projections.
+ * Subclass `CustomPropertyProjection` to introduce new kinds of property projections.
+ */
+
+import javascript
+
+/**
+ * A call that projects selected properties of an object.
```
Comment (xiemaisi) [inline javascript/ql/src/semmle/javascript/frameworks/PropertyProjection.qll]: Here and below: these comments could do with examples of what is being modelled.

File: javascript/ql/src/semmle/javascript/frameworks/PropertyProjection.qll [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+   */
+  abstract DataFlow::Node getObject();
+
+  /**
+   * Gets an argument that selects the properties to project.
+   */
+  abstract DataFlow::Node getASelector();
+
+  /**
+   * Holds if this call returns the value of a single projected property.
```
Comment (xiemaisi) [inline javascript/ql/src/semmle/javascript/frameworks/PropertyProjection.qll]: This is particularly arcane; please explain more carefully and add an example.

**Comment by esben-semmle** [comment]
Comments addressed.

**Comment by esben-semmle** [comment]
Squashed the fixups.

**Comment by xiemaisi** [comment]
The tests actually seem to have succeeded, but the status didn't propagate (@pavel). Merging.

**Comment by xiemaisi** [comment]
@esben-semmle, this PR seems to have been left behind when creating https://github.com/Semmle/ql/tree/rc/1.18. Could you open a PR to cherry-pick the commits across? Same for https://github.com/Semmle/ql/pull/115.

**Comment by pavel** [comment]
@xiemaisi Thanks for notifying me, but it does not seem that I'm involved in this project in any way.

https://blog.github.com/2011-03-23-mention-somebody-they-re-notified/

**Comment by xiemaisi** [comment]
Oops, sorry, wrong Pavel... (It's a great project, though! :smile:)

**Pull request merged**